parquet-column/src/main/java/org/apache/parquet/column/values/bloom/Bloom.java

```diff
@@ -0,0 +1,484 @@
+    this.bloomFilterHash = hash;
+    this.bloomFilterAlgorithm = algorithm;
+  }
+
+  /**
+   * Create a new bitset for bloom filter, at least 256 bits will be create.
+   * @param numBytes number of bytes for bit set.
+   */
+  public void initBitset(int numBytes) {
+    if (numBytes < MINIMUM_BLOOM_SIZE) {
+      numBytes = MINIMUM_BLOOM_SIZE;
+    }
+
+    if (numBytes > ParquetProperties.DEFAULT_MAXIMUM_BLOOM_FILTER_SIZE) {
+      numBytes = ParquetProperties.DEFAULT_MAXIMUM_BLOOM_FILTER_SIZE;
+    }
+
+    // 32 bytes alignment, one bucket.
+    numBytes = (numBytes + BYTES_PER_BUCKET - 1) & ~(BYTES_PER_BUCKET - 1);
+
+    ByteBuffer bytes = ByteBuffer.allocate(numBytes);
+    this.bitset = bytes.array();
+    this.numBytes = numBytes;
+  }
+
+  /**
+   * Create bitset from a given byte array.
+   * @param bitset Given bitset for bloom filter.
+   */
+  public void initBitset(byte[] bitset) {
+    this.bitset = bitset;
+    this.numBytes = bitset.length;
+  }
+
+  /**
+   * @return Bytes in buffered in this bloom filter.
+   */
+  public BytesInput getBytes() {
+    inputs.clear();
+
+    // Add number of bytes.
+    inputs.add(BytesInput.fromInt(numBytes));
+
+    // Add hash strategy
+    inputs.add(BytesInput.fromInt(this.bloomFilterHash.ordinal()));
+
+    // Add bloom filter algorithm
+    inputs.add(BytesInput.fromInt(this.bloomFilterAlgorithm.ordinal()));
+
+    // Add bitset.
+    inputs.add(BytesInput.from(bitset, 0, bitset.length));
+
+    return BytesInput.concat(inputs);
+  }
+
+  private void setMask(int key) {
+    final int SALT[] = {0x47b6137b, 0x44974d91, 0x8824ad5b, 0xa2b7289d,
+      0x705495c7, 0x2df1424b, 0x9efc4947, 0x5c6bfb31};
+
+    Arrays.fill(mask, 0);
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = key * SALT[i];
+    }
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = mask[i] >> 27;
+    }
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = 0x1 << mask[i];
+    }
+
+  }
+
+  /**
+   * Add an element to bloom filter, the element content is represented by
+   * the hash value of its plain encoding result.
+   * @param hash hash result of element.
+   */
+  public void bloomInsert(long hash) {
+    int bucketIndex = (int)(hash >> 32) & (numBytes / BYTES_PER_BUCKET - 1);
+    int key = (int)hash;
+
+    setMask(key);
+
+    for (int i = 0; i < 8; i++) {
+      int bitsetIndex = bucketIndex * BYTES_PER_BUCKET + i * 4;
+      bitset[bitsetIndex] |= (byte)(mask[i] >>> 24);
+      bitset[bitsetIndex + 1] |= (byte)(mask[i] >>> 16);
+      bitset[bitsetIndex + 2] |= (byte)(mask[i] >>> 8);
+      bitset[bitsetIndex + 3] |= (byte)(mask[i]);
+    }
+  }
+
+  /**
+   * Determine where an element is in set or not.
+   * @param hash the hash value of element plain encoding result.
+   * @return false if element is must not in set, true if element probably in set.
+   */
+  public boolean bloomFind(long hash) {
+    int bucketIndex = (int)(hash >> 32) & (numBytes / BYTES_PER_BUCKET - 1);
```

Comment:
This returns incorrect results when elements is empty.

Reply:
when elements is empty, it return false since no bit is set in bitset. Why it returns wrong result?